server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentCompactor.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.coordinator.helper;
+
+import com.google.inject.Inject;
+import io.druid.client.indexing.IndexingServiceClient;
+import io.druid.client.indexing.QueryStatus;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.server.coordinator.CoordinatorCompactionConfig;
+import io.druid.server.coordinator.CoordinatorStats;
+import io.druid.server.coordinator.DruidCoordinatorRuntimeParams;
+import io.druid.timeline.DataSegment;
+import io.druid.timeline.VersionedIntervalTimeline;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class DruidCoordinatorSegmentCompactor implements DruidCoordinatorHelper
+{
+  static final String COMPACT_TASK_COUNT = "compactTaskCount";
+  static final String SEGMENTS_WAIT_COMPACT = "segmentsWaitCompact";
+
+  private static final Logger LOG = new Logger(DruidCoordinatorSegmentCompactor.class);
+
+  private final CompactionSegmentSearchPolicy policy = new NewestSegmentFirstPolicy();
+  private final IndexingServiceClient indexingServiceClient;
+  private final Set<String> queryIds = new HashSet<>();
```

Comment:
I don't see this getting initialized anywhere when coordinator starts .. e.g. on a leadership change , new coordinator would not take compaction tasks into account which were started by previous leader.

Reply:
I think the word query is used because of the current naming in `IndexingServiceClient` , it might still make sense to use word `task` instead in various logs and variables in this class e.g. instead of queryIds using compationTaskIds .

Reply:
> I don't see this getting initialized anywhere when coordinator starts .. e.g. on a leadership change , new coordinator would not take compaction tasks into account which were started by previous leader.

Good point! I'm thinking to get the number of running compaction tasks when ```DruidCoordinatorSegmentCompactor.run()``` is called, and I'll add it soon.

BTW, the refactoring done in https://github.com/druid-io/druid/pull/5149 will be helpful to implement it. Some classes are added or moved to ```druid-api``` to be reused by overlord clients without code duplication. It will be easy to get the number of running compaction tasks this once https://github.com/druid-io/druid/pull/5149 is merged.

Reply:
> I think the word query is used because of the current naming in IndexingServiceClient , it might still make sense to use word task instead in various logs and variables in this class e.g. instead of queryIds using compationTaskIds .

Ok. I'll rename it together.